solr/solrj/src/test/org/apache/solr/client/solrj/impl/LBHttp2SolrClientTest.java

```diff
@@ -51,4 +67,249 @@ public void testLBHttp2SolrClientWithTheseParamNamesInTheUrl() {
+
+      int limit = 10; // For simplicity use an even limit
+      int halfLimit = limit / 2; // see TODO below
+
+      CountDownLatch cdl = new CountDownLatch(limit); // deprecated API use
+      List<LBTestAsyncListener> listeners = new ArrayList<>(); // deprecated API use
+      List<CompletableFuture<LBSolrClient.Rsp>> responses = new ArrayList<>();
+
+      for (int i = 0; i < limit; i++) {
+        QueryRequest queryRequest = new QueryRequest(new MapSolrParams(Map.of("q", "" + i)));
+        LBSolrClient.Req req = new LBSolrClient.Req(queryRequest, endpointList);
+        if (useDeprecatedApi) {
+          LBTestAsyncListener listener = new LBTestAsyncListener(cdl);
+          listeners.add(listener);
+          testClient.asyncReq(req, listener);
+        } else {
+          responses.add(testClient.requestAsync(req));
+        }
+      }
+
+      if (useDeprecatedApi) {
+        try {
+          // This is just a formality.  This is a single-threaded test.
+          cdl.await(1, TimeUnit.MINUTES);
+        } catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          fail("interrupted");
+        }
+      }
+
+      QueryRequest[] queryRequests = new QueryRequest[limit];
+      int numEndpointOne = 0;
+      int numEndpointTwo = 0; // see TODO below
+      for (int i = 0; i < limit; i++) {
+        SolrRequest<?> lastSolrReq = client.lastSolrRequests.get(i);
+        assertTrue(lastSolrReq instanceof QueryRequest);
+        QueryRequest lastQueryReq = (QueryRequest) lastSolrReq;
+        int index = Integer.parseInt(lastQueryReq.getParams().get("q"));
+        assertNull("Found same request twice: " + index, queryRequests[index]);
+        queryRequests[index] = lastQueryReq;
+        if (lastQueryReq.getBasePath().equals(ep1.toString())) {
+          numEndpointOne++;
+        } else if (lastQueryReq.getBasePath().equals(ep2.toString())) {
+          numEndpointTwo++;
+        }
+        NamedList<Object> lastResponse;
+        if (useDeprecatedApi) {
+          LBTestAsyncListener lastAsyncListener = listeners.get(index);
+          assertTrue(lastAsyncListener.onStartCalled);
+          assertNull(lastAsyncListener.failure);
+          assertNotNull(lastAsyncListener.success);
+          lastResponse = lastAsyncListener.success.getResponse();
+        } else {
+          LBSolrClient.Rsp lastRsp = null;
+          try {
+            lastRsp = responses.get(index).get();
+          } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            fail("interrupted");
+          } catch (ExecutionException ee) {
+            fail("Response " + index + " ended in failure: " + ee);
+          }
+          lastResponse = lastRsp.getResponse();
+        }
+
+        // The Mock will return {"response": index}.
+        assertEquals("" + index, lastResponse.get("response"));
+      }
+
+      // TODO: LBHttp2SolrClient creates a new "endpoint iterator" per request, thus
```

Comment:
Huh.  I think they should behave the same, first of all, because it's not pertinent to the HTTP Client.  But I think the first is supposed to be chosen because the order can be chosen intentionally (say prefer the replica that is a leader, or prefer pull replicas, etc.) and passed in sorted by such rules.

Reply:
My concern is that unless the first node in the passed-in list is in the Zombies, it always gets all the requests.  Maybe this is the intention, that the caller is supposed to shuffle the list themselves so as to give ultimate control.  But this is not what I expected from a class named "Load Balancing Client".  There seemed to be little by way of doc comments and unit tests to confirm this. Probably, we need a follow-up issue to add more test coverage to these LB clients and then better document or fix the behavior.

Reply:
Understood.
BTW to be more specific, the sorting happens here: https://github.com/apache/solr/blob/c5709bd4135538fffba3b2c4947dd91ff28d5750/solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java#L1018 with ReplicaListTransformers that sort & shuffle.

Reply:
Ok, I can try and improve the doc comment to let the user know they need to sort the list themselves and also take the TODO's out of the unit test.  But as someone who isn't expert on the codebase, I was very surprised when I found it does not actually Load Balance.  It seems there is too tight of coupling between Cloud Solr Client and LB Solr Client here.